marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java

```diff
@@ -1,477 +1,187 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicLong;
-import java.util.stream.Collectors;
-
-public class FlowRunnerImpl implements FlowRunner {
-
-    private static final int DEFAULT_BATCH_SIZE = 100;
-    private static final int DEFAULT_THREAD_COUNT = 4;
-    private static final int MAX_ERROR_MESSAGES = 10;
-    private Flow flow;
-    private int batchSize = DEFAULT_BATCH_SIZE;
-    private int threadCount = DEFAULT_THREAD_COUNT;
-    private DatabaseClient stagingClient;
-    private String destinationDatabase;
-    private Map<String, Object> options;
-    private int previousPercentComplete;
-    private boolean stopOnFailure = false;
-    private String jobId;
-    private boolean isFullOutput = false;
-
-
-    private int step = 1;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
 
-    private List<FlowItemCompleteListener> flowItemCompleteListeners = new ArrayList<>();
-    private List<FlowItemFailureListener> flowItemFailureListeners = new ArrayList<>();
-    private List<FlowStatusListener> flowStatusListeners = new ArrayList<>();
-    private List<FlowFinishedListener> flowFinishedListeners = new ArrayList<>();
-
-    private HubConfig hubConfig;
-    private Thread runningThread = null;
-    private DataMovementManager dataMovementManager = null;
-    private QueryBatcher queryBatcher = null;
-    private JobUpdate jobUpdate = new JobUpdate(hubConfig.newJobDbClient());
+@Component
+public class FlowRunnerImpl implements FlowRunner{
 
-    public FlowRunnerImpl(HubConfig hubConfig) {
-        this.hubConfig = hubConfig;
-        this.stagingClient = hubConfig.newStagingClient();
-        this.destinationDatabase = hubConfig.getDbName(DatabaseKind.FINAL);
-    }
+    @Autowired
+    private HubConfigImpl hubConfig;
 
-    public FlowRunner withFlow(Flow flow) {
-        this.flow = flow;
-        return this;
-    }
+    @Autowired
+    private FlowManager flowManager;
 
-    public FlowRunner withStep(int step) {
-        this.step = step;
-        return this;
-    }
+    private AtomicBoolean isRunning;
+    private AtomicBoolean isJobCancelled;
 
-    public FlowRunner withJobId(String jobId) {
-        this.jobId = jobId;
-        return this;
-    }
+    private String runningJobId;
+    private Step runningStep;
+    private Flow runningFlow;
 
-    @Override
-    public FlowRunner withBatchSize(int batchSize) {
-        this.batchSize = batchSize;
-        return this;
-    }
+    private StepRunner stepRunner;
 
-    @Override
-    public FlowRunner withThreadCount(int threadCount) {
-        this.threadCount = threadCount;
-        return this;
-    }
+    private Map<String, Queue<String>> stepsMap = new HashMap<>();
+    private Map<String, Flow> flowMap = new HashMap<>();
+    private Map<String, RunFlowResponse> flowResp = new HashMap<>();
```

Comment:
https://github.com/marklogic/marklogic-data-hub/blob/522cf42f47a9fb436e7727eaa57a677d5db930e7/marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java#L44-L46

We might need to use ConcurrentHashMap instead of HashMap for thread safety...

Reply:
I will make this change.. thx for the suggestion